### PR TITLE
Add support for sndio

### PR DESCRIPTION
**Pull request by brad0**
This is based on current modifications to portaudio done in OpenBSD.

Closes: #220

### CODE REVIEW

**Comment by philburk** [comment]
@brad0 - Thanks for taking this on. This PR can replace #647

I notice that the yaml CI is not being run! I wonder why.

**Comment by brad0** [comment]
> @brad0 - Thanks for taking this on. This PR can replace #647
> 
> I notice that the yaml CI is not being run! I wonder why.

Yup. I am continuing off of 647 plus the few suggestions made that were left.

My last push was to also re-generate the autoconf script since I noticed it was being used during CI without it's own regeneration. I see it wasn't update since Pulse was added either.

File: src/hostapi/sndio/pa_sndio.c [hunk 1]
```diff
@@ -0,0 +1,763 @@
...
+    const char *dev;
+
+    PA_DEBUG( ( "OpenStream:\n" ) );
+
+    mode = 0;
+    inputChannelCount = outputChannelCount = 0;
+    ifmt = ofmt = 0;
+    sio_initpar( &par );
+
+    if( outputPar && outputPar->channelCount > 0 )
```
Comment (philburk) [inline src/hostapi/sndio/pa_sndio.c:267]: I'm stumped. Is this an outputParticle, outputParagraph, outputParallel? ;-)

File: src/hostapi/sndio/pa_sndio.c [hunk 2]
```diff
@@ -0,0 +1,763 @@
...
+                           void *userData )
+{
+    PaSndioHostApiRepresentation *sndioHostApi = (PaSndioHostApiRepresentation *)hostApi;
+    PaSndioStream *sndioStream;
+    PaError err;
+    struct sio_hdl *hdl;
+    struct sio_par par;
+    unsigned mode;
+    int inputChannelCount, outputChannelCount;
+    PaSampleFormat ifmt, ofmt, siofmt;
```
Comment (philburk) [inline src/hostapi/sndio/pa_sndio.c]: Please use more clear names like "inputFormat".

**Review by philburk — CHANGES_REQUESTED** [review]
Great. I notice almost all of the comments from #647 have been resolved.
There are just a few cryptic variable names.
We are very close to getting this merged.

File: src/hostapi/sndio/pa_sndio.c (hunk 1 above)
Comment (brad0) [reply src/hostapi/sndio/pa_sndio.c:267]: I believe to be Parameters in this context.

Comment (philburk) [reply src/hostapi/sndio/pa_sndio.c:267]: Ah! Obvious in hindsight. I was honestly stumped. Thanks for the clarification.

**Comment by philburk** [comment]
@brad0 - I really appreciate you taking this on. Ross and I were looking at old stale PRs
and this one looked very close to being ready. Just a few tweaks and we will be able to add sndio support. Yay!

**Comment by brad0** [comment]
> @brad0 - I really appreciate you taking this on. Ross and I were looking at old stale PRs and this one looked very close to being ready. Just a few tweaks and we will be able to add sndio support. Yay!

Well I would like to help where I can to get this in. As I said in the other PR we have had a PA backend in our ports tree for 15 years. So it would be good to finally have something go upstream.

**Comment by brad0** [comment]
Is there anything else you would like to see changed?

**Comment by brad0** [comment]
@philburk Ping.

File: src/hostapi/sndio/pa_sndio.c [hunk 3]
```diff
@@ -0,0 +1,769 @@
...
+     * Normally we just expose the default device, which can be
+     * changed via the AUDIODEVICE environment variable, but we
+     * also allow specifying a list of up to 16 devices via the
+     * PA_SNDIO_AUDIODEVICES environment variable.
+     *
+     * Example:
+     * PA_SNDIO_AUDIODEVICES=default:snd/0.monitor:snd@remote/0
+     */
+#define PA_SNDIO_AUDIODEVICES_MAX 16
+    PaDeviceInfo device_info[PA_SNDIO_AUDIODEVICES_MAX];
```
Comment (RossBencina) [inline src/hostapi/sndio/pa_sndio.c:72]: `device_info` -> `deviceInfos`

File: src/hostapi/sndio/pa_sndio.c [hunk 4]
```diff
@@ -0,0 +1,769 @@
...
+    return 1;
+}
+
+/*
+ * I/O loop for callback interface
+ */
+static void *sndioThread( void *arg )
+{
+    PaSndioStream *sndioStream = (PaSndioStream *)arg;
+    PaStreamCallbackTimeInfo timeinfo;
```
Comment (RossBencina) [inline src/hostapi/sndio/pa_sndio.c:168]: `timeinfo` -> `timeInfo`

File: src/hostapi/sndio/pa_sndio.c [hunk 5]
```diff
@@ -0,0 +1,769 @@
...
+                           void *userData )
+{
+    PaSndioHostApiRepresentation *sndioHostApi = (PaSndioHostApiRepresentation *)hostApi;
+    PaSndioStream *sndioStream;
+    PaError err;
+    struct sio_hdl *hdl;
+    struct sio_par par;
+    unsigned mode;
+    int inputChannelCount, outputChannelCount;
+    PaSampleFormat inputFormat, outputFormat, SndioFormat;
```
Comment (RossBencina) [inline src/hostapi/sndio/pa_sndio.c:257]: `SndioFormat` -> `sndioFormat`

File: src/hostapi/sndio/pa_sndio.c [hunk 6]
```diff
@@ -0,0 +1,769 @@
...
+     * changed via the AUDIODEVICE environment variable, but we
+     * also allow specifying a list of up to 16 devices via the
+     * PA_SNDIO_AUDIODEVICES environment variable.
+     *
+     * Example:
+     * PA_SNDIO_AUDIODEVICES=default:snd/0.monitor:snd@remote/0
+     */
+#define PA_SNDIO_AUDIODEVICES_MAX 16
+    PaDeviceInfo device_info[PA_SNDIO_AUDIODEVICES_MAX];
+    PaDeviceInfo *infos[PA_SNDIO_AUDIODEVICES_MAX];
```
Comment (RossBencina) [inline src/hostapi/sndio/pa_sndio.c:73]: `infos` -> `deviceInfoPtrs`

**Review by RossBencina — APPROVED** [review]
I don't want to stand in the way of merging this, but here are a few variable namings that would be nice to  clean up.

**Comment by brad0** [comment]
> I don't want to stand in the way of merging this, but here are a few variable namings that would be nice to clean up.

I don't make making changes. I just need specific feedback.

**Review by philburk — APPROVED** [review]

**Comment by RossBencina** [comment]
Thanks @brad0. We're keen to accept any fixes, updates or cleanup as they become available.

**Comment by brad0** [comment]
> Thanks @brad0. We're keen to accept any fixes, updates or cleanup as they become available.

Thank you. If anything does come up I'll make sure to submit any such changes.

**Pull request merged**